runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -114,7 +144,528 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
+					throwOOM = true;
+					goto done;
+				}
+				 /* Push the object on the special frame so as to avoid updating the address on the stack
+				  * by GC when allocating memory for the next pointer/struct of the argument list.
+				  */
+				PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, memAddrObject);
+				break;
+			}
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+			{
+				I_64 offset = (I_64)(intptr_t)vmFuncs->getArgPointer(nativeSig, argsListPointer, argIndex);
+				j9object_t memSegmtObject = createMemSegmentObject(data, offset, sigArray[argIndex].sizeInByte, isCurThrdAllocated, sessionOrScopeObject);
+				if (NULL == memSegmtObject) {
+					/* The OOM exception set in createMemSegmentObject() will be thrown in the interpreter
+					 * after returning from the native function in downcall.
+					 */
+					throwOOM = true;
+					goto done;
+				}
+				 /* Push the object on the special frame so as to avoid updating the address on the stack
+				  * by GC when allocating memory for the next pointer/struct of the argument list.
+				  */
+				PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, memSegmtObject);
+				break;
+			}
+			default:
+				Assert_VM_unreachable();
+				break;
+			}
+		}
+		/* Only restore the literals and keep the unchanged arguments on the stack which
+		 * are passed over to native2InterpreterTransition() in the interpreter so as to
+		 * invoke the upcall method handle.
+		 */
+		currentThread->literals = thrLiterals;
+
+		/* Place mhMetaData as the return value to native2InterpreterTransition() when calling into
+		 * the interpreter so as to set the invoke cache array (MemberName and appendix)
+		 * before invoking the target handle in upcall.
+		 */
+		currentThread->returnValue = J9_BCLOOP_N2I_TRANSITION;
+		currentThread->returnValue2 = (UDATA)data;
+		c_cInterpreter(currentThread);
+
+done:
+		restoreCallInFrameHelper(currentThread);
+	}
+	VM_VMAccess::inlineExitVMToJNI(currentThread);
+
+	if (!throwOOM) {
+		returnStorage = currentThread->returnValue;
+		convertUpcallReturnValue(data, returnType, &returnStorage);
+	}
+
+	/* Release the J9VMThread if allocated locally for the native thread */
+	if (isCurThrdAllocated) {
+		threadCleanup(currentThread, false);
+		currentThread = NULL;
+	}
+
+doneAndExit:
+	return returnStorage;
+}
+
+/**
+ * @brief Get a J9VMThread whether it is the current thread or a newly created thread if doesn't exist
+ *
+ * Note:
+ * The function is to handle the situation when there is no Java thread for the current native thread
+ * which is directly created in native code without a Java thread attached to it.
+ *
+ * @param vm a pointer to J9UpcallMetaData
+ * @param isCurThrdAllocated a pointer to a flag indicating whether the current thread is allocated locally
+ * @return a pointer to J9VMThread
+ */
+static J9VMThread *
+getCurrentThread(J9UpcallMetaData *data, bool *isCurThrdAllocated)
+{
+	J9JavaVM *vm = data->vm;
+	J9VMThread *currentThread = vm->internalVMFunctions->currentVMThread(vm);
```

Comment:
same here

Reply:
Agreed and removed `vm->internalVMFunctions->` at multiple places.